packages/data-point/lib/entity-types/base-entity/resolve.js

```diff
@@ -139,26 +139,30 @@ function resolveEntity (
   const resolveReducerBound = _.partial(resolveReducer, manager)
 
   return Promise.resolve(accUid)
-    .then(acc => resolveMiddleware(manager, `before`, acc))
     .then(acc =>
-      resolveMiddleware(manager, `${reducer.entityType}:before`, acc)
+      typeCheck(manager, acc, acc.reducer.spec.inputType, resolveReducer)
     )
+    .then(acc => resolveMiddleware(manager, `before`, acc))
     .then(acc =>
-      typeCheck(manager, acc, acc.reducer.spec.inputType, resolveReducer)
+      resolveMiddleware(manager, `${reducer.entityType}:before`, acc)
     )
     .then(acc => resolveReducer(manager, acc, acc.reducer.spec.before))
     .then(acc => mainResolver(acc, resolveReducerBound))
     .then(acc => resolveReducer(manager, acc, acc.reducer.spec.after))
-    .then(acc =>
-      typeCheck(manager, acc, acc.reducer.spec.outputType, resolveReducer)
-    )
     .then(acc => resolveMiddleware(manager, `${reducer.entityType}:after`, acc))
     .then(acc => resolveMiddleware(manager, `after`, acc))
     .catch(error => {
       // checking if this is an error to bypass the `then` chain
       if (error.bypass === true) {
         return error.bypassValue
       }
+
+      throw error
+    })
+    .then(acc =>
+      typeCheck(manager, acc, acc.reducer.spec.outputType, resolveReducer)
+    )
```

Comment:
this still doesn't type check the result of the `error` lifecycle method, because that's resolved using `resolveErrorReducers` in the following catch block (and after we this, we should add a test where the `error` reducer returns an invalid type)

Reply:
I’m not sure I understand - the `error` lifecycle method should conform to the validation done by `outputType`?

Reply:
If somebody uses `error` to return a default value, that value would not be type checked:

```js
const dataPoint = DataPoint.create({
  entities: {
    'model:x': {
      value: () => {
        throw new Error()
      },
      error: () => 'STRING',
      outputType: 'number'
    }
  }
})

dataPoint.resolve('model:x', 1) // => 'STRING' (but outputType is 'number')
```

Reply:
That seems kinda reasonable though. What if the entity is a `request` that normally returns an array, the `error` shouldn’t throw an error object?

Also, I don’t think your example would get caught as an error.

```js
dataPoint.resolve('model:x', 1)
  .then(val => {
    // val === 'STRING'
  })
  .catch(e => {
    // no error thrown
  })
```

Maybe we need to account for when the `error` returns instead of throwing?